src/builder/QuarkBuilder.sol

```diff
@@ -311,52 +475,36 @@ contract QuarkBuilder {
-        assertTransferActionsAffordableOnTargetChain(
-            transferActions, chainAccountsList, transferIntent.chainId, plannedBridgeAmount
-        );
-    }
-
-    function assertTransferActionsAffordableOnTargetChain(
-        Actions.Action[] memory transferActions,
-        Accounts.ChainAccounts[] memory chainAccountsList,
-        uint256 targetChainId,
-        uint256 plannedBridgeAmountToTargetChain
-    ) internal pure {
-        uint256 paymentTokensUsed = 0;
-        for (uint256 i = 0; i < transferActions.length; ++i) {
-            Actions.TransferActionContext memory transferActionContext =
-                abi.decode(transferActions[i].actionContext, (Actions.TransferActionContext));
-            // Filter with the targetChainId and paymentTokensUsed will track on one chain at a time
-            if (transferActionContext.chainId == targetChainId) {
-                address transferToken = transferActionContext.token;
-                uint256 transferAmount = transferActionContext.amount;
-                uint256 paymentAssetBalanceOnChain = Accounts.sumBalances(
-                    Accounts.findAssetPositions(transferToken, transferActions[i].chainId, chainAccountsList)
-                );
-                paymentTokensUsed += transferActions[i].paymentMaxCost;
-                if (transferToken == transferActions[i].paymentToken) {
-                    // If the payment token is the transfer token and this is the target chain, we need to account for the transfer amount
-                    // If its transfer step, check if user has enough balance to cover the transfer amount after bridge
-                    if (
-                        paymentAssetBalanceOnChain + plannedBridgeAmountToTargetChain
-                            < paymentTokensUsed + transferAmount
-                    ) {
-                        revert MaxCostTooHigh();
-                    }
-
-                    // Special handling as the payment token is sent out so it will be part of the cost
-                    paymentTokensUsed += transferAmount;
-                } else {
-                    // Just check payment token can cover the max cost
-                    if (paymentAssetBalanceOnChain < paymentTokensUsed) {
-                        revert MaxCostTooHigh();
-                    }
+        Actions.Action[] memory nonBridgeActions = Actions.findActionsNotOfType(actions, Actions.ACTION_TYPE_BRIDGE);
+        address paymentToken = nonBridgeActions[0].paymentToken; // assumes all actions use the same paymentToken
+        uint256 targetChainPaymentTokenBalance =
+            Accounts.sumBalances(Accounts.findAssetPositions(paymentToken, targetChainId, chainAccountsList)); // assumes that all non-bridge actions occur on the same chain
+        uint256 paymentTokenCost = 0;
+
+        for (uint256 i = 0; i < nonBridgeActions.length; ++i) {
+            // revert if not on the target chain?
```

Comment:
Wondering if we should revert if there is a non-bridge action that is not on the destination chain.

Reply:
Sure, since we make that assumption, it's good to revert if that assumption is broken.

Reply:
added